src/lib.rs

```diff
@@ -242,25 +257,33 @@ macro_rules! unsafe_abomonate {
 
 // general code for tuples (can't use '0', '1', ... as field identifiers)
 macro_rules! tuple_abomonate {
-    ( $($name:ident)+) => (
-        impl<$($name: Abomonation),*> Abomonation for ($($name,)*) {
+    ( $($ty:ident)+) => (
+        impl<$($ty: Abomonation),*> Abomonation for ($($ty,)*) {
             #[allow(non_snake_case)]
             #[inline(always)] unsafe fn entomb<WRITE: Write>(&self, write: &mut WRITE) -> IOResult<()> {
-                let ($(ref $name,)*) = *self;
-                $($name.entomb(write)?;)*
+                let ($(ref $ty,)*) = *self;
+                $($ty.entomb(write)?;)*
                 Ok(())
             }
+
             #[allow(non_snake_case)]
-            #[inline(always)] unsafe fn exhume<'a,'b>(&'a mut self, mut bytes: &'b mut [u8]) -> Option<&'b mut [u8]> {
-                let ($(ref mut $name,)*) = *self;
-                $( let temp = bytes; bytes = $name.exhume(temp)?; )*
+            #[inline(always)] unsafe fn exhume<'a>(mut self_: NonNull<Self>, mut bytes: &'a mut [u8]) -> Option<&'a mut [u8]> {
+                // FIXME: This (briefly) constructs a "ref mut" to invalid data, which is UB.
+                //        I think avoiding this would require a cleaner way to iterate over tuple fields.
+                //        One possibility would be a C++11-style combination of variadic generics and recursion.
+                let ($(ref mut $ty,)*) = *self_.as_ptr();
+                $(
+                    let field_ptr : NonNull<$ty> = From::from($ty);
```

Comment:
My inner programming language nerd is amazed by the fact that Rust's grammar is so well designed that we can have a type and a variable with the same name and use these at the same time without any issues.

I also wish we didn't need this hack